ffsync/ordered_id.go

```diff
@@ -83,3 +93,169 @@ func (m *memoryIdGenerator) NextId(namespace string) (OrderedId, error) {
+	session *concurrency.Session
+	ctx     context.Context
+}
+
+func (etcd *etcdIdGenerator) NextId(namespace string) (OrderedId, error) {
+	if namespace == "" {
+		return nil, fferr.NewInternalError(fmt.Errorf("cannot generate ID for empty namespace"))
+	}
+
+	// Lock the namespace to prevent concurrent ID generation
+	lockKey := etcd_id_key + namespace
+	lockMutex := concurrency.NewMutex(etcd.session, lockKey)
+	if err := lockMutex.Lock(etcd.ctx); err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to lock key %s: %w", lockKey, err))
+	}
+	defer lockMutex.Unlock(etcd.ctx)
+
+	// Get the current ID
+	resp, err := etcd.client.Get(etcd.ctx, lockKey)
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to get key %s: %w", lockKey, err))
+	}
+
+	// Initialize the ID if it doesn't exist
+	var nextId uint64
+	if len(resp.Kvs) == 0 {
+		nextId = 1
+	} else {
+		if err := json.Unmarshal(resp.Kvs[0].Value, &nextId); err != nil {
+			return nil, fferr.NewInternalError(fmt.Errorf("failed to unmarshal ID: %w", err))
+		}
+		nextId++
+	}
+
+	// Update the ID
+	nextIdBytes, err := json.Marshal(nextId)
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to marshal ID: %w", err))
+	}
+	if _, err := etcd.client.Put(etcd.ctx, lockKey, string(nextIdBytes)); err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to set key %s: %w", lockKey, err))
+	}
+
+	return (*Uint64OrderedId)(&nextId), nil
+}
+
+func (etcd *etcdIdGenerator) Close() {
+	etcd.session.Close()
+	etcd.client.Close()
+}
+
+func NewRDSOrderedIdGenerator(config helpers.RDSConfig) (OrderedIdGenerator, error) {
+	const tableName = "ff_ordered_id"
+
+	db, err := helpers.NewRDSPoolConnection(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// acquire a connection pool
+	connection, err := db.Acquire(context.Background())
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to acquire connection from the database pool: %w", err))
+	}
+
+	// ping the database
+	err = connection.Ping(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("failed to ping the database: %w", err)
+	}
+
+	// Create the id table if it doesn't exist
+	tableCreationSQL := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s (namespace VARCHAR(2048) PRIMARY KEY, current_id BIGINT)", helpers.SanitizePostgres(tableName))
+	_, err = db.Exec(context.Background(), tableCreationSQL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create table %s: %w", tableName, err)
+	}
+
+	return &rdsIdGenerator{
+		db:         db,
+		tableName:  tableName,
+		connection: connection,
+	}, nil
+}
+
+type rdsIdGenerator struct {
+	db         *pgxpool.Pool
+	tableName  string
+	connection *pgxpool.Conn
+}
+
+func (rds *rdsIdGenerator) NextId(namespace string) (OrderedId, error) {
+	if namespace == "" {
+		return nil, fmt.Errorf("cannot generate ID for empty namespace")
+	}
+
+	tx, err := rds.db.BeginTx(context.Background(), pgx.TxOptions{})
```

Comment:
Turn the query into a single one using an upsert then you don't need a transaction